instrumentation/servlet/src/main/java/brave/servlet/internal/ServletRuntime.java

```diff
@@ -91,43 +92,55 @@ private static ServletRuntime findServletRuntime() {
     }
 
     @Override public void handleAsync(
-      HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler,
+      HttpServerHandler<HttpServerRequest, HttpServerResponse> handler,
       HttpServletRequest request, HttpServletResponse response, Span span) {
       if (span.isNoop()) return; // don't add overhead when we aren't httpTracing
       TracingAsyncListener listener = new TracingAsyncListener(handler, span);
       request.getAsyncContext().addListener(listener, request, response);
     }
 
     static final class TracingAsyncListener implements AsyncListener {
-      final HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler;
+      final HttpServerHandler<HttpServerRequest, HttpServerResponse> handler;
       final Span span;
-      volatile boolean complete; // multiple async events can occur, only complete once
+      final AtomicBoolean complete = new AtomicBoolean(); // multiple async events can occur
 
-      TracingAsyncListener(
-        HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler,
+      TracingAsyncListener(HttpServerHandler<HttpServerRequest, HttpServerResponse> handler,
         Span span
       ) {
         this.handler = handler;
         this.span = span;
       }
 
       @Override public void onComplete(AsyncEvent e) {
-        if (complete) return;
-        handler.handleSend(httpServerResponse(e), null, span);
-        complete = true;
+        if (!complete.compareAndSet(false, true)) {
+          // TODO: None of our tests reach this condition. Make a concrete case that re-enters the
```

Comment:
@nicmunroe @felixbarny PS I did something special to execute the entire suite and fail if this ever was the case.. it never was. Right now, I know I've been told this could be re-entrant, but I'd like to see what test we are missing. Or if any of you have a test that proves it is re-entrant?

Reply:
@adriancole I can't remember if I've ever seen it be reentrant in practice. But it's servlet, so my cynical side says it's always possible. Even if you find something in the servlet spec that says impls shouldn't do it, I bet there's some framework that does either due to a bug or because they didn't realize they shouldn't or they feel like they have a use case that warrants it. That's all FUD - I don't have any concrete examples. But I wouldn't be surprised in the least. The question is whether it's worth the extra bulletproofing if we can't find a way to reproduce it. I don't have a good answer on that one.

So I don't really have any advice on this one, sorry. I think there are defensible arguments for going either way.

Reply:
Thanks for the feedback Nic! in our case, the only impact to doing this twice is overhead, because this is on the finish hook not the start hook (eg it won't leak a span). In this case, and as I can't reproduce it normally, I'm inclined to remove the bullet-proofing, maybe mentioning that there's no impact to calling twice (and a test to prove it)